File: 1470.shuffle-the-array.java
/*
 * @lc app=leetcode id=1470 lang=java
 *
 * [1470] Shuffle the Array
 */

// @lc code=start
class Solution {
    public int[] shuffle(int[] nums, int n) {
        final int mask = 0x03ff;
        for (int ix = 0; ix != n; ++ix) {
            nums[ix * 2] = nums[ix * 2] | ((nums[ix] & mask) << 10);
            nums[ix * 2 + 1] = nums[ix * 2 + 1] | ((nums[ix + n] & mask) << 10);
        }
        for (int ix = 0; ix != nums.length; ++ix) {
            nums[ix] = (nums[ix] >> 10) & mask;
        }
        return nums;
    }
}
// @lc code=end

File: 1470.shuffle-the-array.cpp
/*
 * @lc app=leetcode id=1470 lang=cpp
 *
 * [1470] Shuffle the Array
 */

// @lc code=start
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> shuffle(vector<int>& nums, int n) {
        vector<int> ret(n * 2); auto iter = ret.begin();
        for (auto i = nums.begin(), j = nums.begin() + n; i != nums.begin() + n; ++i, ++j) {
            *(iter++) = *i; *(iter++) = *j;
        }
        return ret;
    }
};
// @lc code=end
